src/agentscope/service/multi_modality/openai_services.py

```diff
@@ -0,0 +1,536 @@
+    except Exception as e:
+        return ServiceResponse(
+            ServiceExecStatus.ERROR,
+            str(e),
+        )
+
+
+def openai_edit_image(
+    image_url: str,
+    mask_url: str,
+    prompt: str,
+    api_key: str,
+    n: int = 1,
+    size: Literal[
+        "256x256",
+        "512x512",
+        "1024x1024",
+    ] = "256x256",
+    save_dir: Optional[str] = None,
+) -> ServiceResponse:
+    """
+    Edit an image based on the provided mask and prompt, and return the edited
+    image URL(s) or save them locally.
+
+    Args:
+        image_url (`str`):
+            The file path or URL to the image that needs editing.
+        mask_url (`str`):
+            The file path or URL to the mask image that specifies the regions
+            to be edited.
+        prompt (`str`):
+            The text prompt describing the edits to be made to the image.
+        api_key (`str`):
+            The API key for the OpenAI API.
+        n (`int`, defaults to `1`):
+            The number of edited images to generate.
+        size (`Literal["256x256", "512x512", "1024x1024"]`, defaults to
+        `"256x256"`):
+            The size of the edited images.
+        save_dir (`Optional[str]`, defaults to `None`):
+            The directory to save the edited images. If not specified, will
+            return the web URLs.
+
+    Returns:
+        `ServiceResponse`:
+            A dictionary with two variables: `status` and `content`.
+            If `status` is `ServiceExecStatus.SUCCESS`,
+            the `content` is a dict with key 'image_urls' and
+            value is a list of the paths to the edited images or URLs.
+
+    Example:
+
+        .. code-block:: python
+
+            image_url = "/path/to/original_image.png"
+            mask_url = "/path/to/mask_image.png"
+            prompt = "Add a sun to the sky"
+            api_key = "YOUR_API_KEY"
+            print(openai_edit_image(image_url, mask_url, prompt, api_key))
+
+        > {
+        >     'status': 'SUCCESS',
+        >     'content': {'image_urls': ['EDITED_IMAGE_URL1',
+        'EDITED_IMAGE_URL2']}
+        > }
+    """
+    os.environ["OPENAI_API_KEY"] = api_key
+    client = OpenAI()
+    # convert relative path to absolute path
+    image_url = os.path.abspath(image_url)
+    mask_url = os.path.abspath(mask_url)
+    with open(image_url, "rb") as img, open(mask_url, "rb") as mask:
+        response = client.images.edit(
+            model="dall-e-2",
+            image=img,
+            mask=mask,
+            prompt=prompt,
+            n=n,
+            size=size,
+        )
+    try:
+        urls = _handle_openai_img_response(response, save_dir)
+        return ServiceResponse(
+            ServiceExecStatus.SUCCESS,
+            {"image_urls": urls},
+        )
+    except Exception as e:
+        return ServiceResponse(
+            ServiceExecStatus.ERROR,
+            str(e),
+        )
+
+
+def openai_create_image_variation(
+    image_url: str,
```

Comment:
So here we support both local and web URLs?

Reply:
Yes.  As `agentscope.models.openai_model` does not provide wrappers for editing and creating image variation, I use `_parse_url` function that takes either local or web URLs as input, and return objects `BytesIO` .